ambry-router/src/test/java/com.github.ambry.router/CryptoJobHandlerTest.java

```diff
@@ -150,16 +172,30 @@ public void testEncryptionFailure() throws InterruptedException, GeneralSecurity
   public void testDecryptionFailure() throws InterruptedException, GeneralSecurityException {
     cryptoJobHandler.close();
     MockCryptoService mockCryptoService = new MockCryptoService(new CryptoServiceConfig(verifiableProperties));
+    IllegalStateException illegalStateException = new IllegalStateException();
     mockCryptoService.exceptionOnDecryption.set(
-        new GeneralSecurityException("Exception to test", new IllegalStateException()));
-    cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT);
+        new GeneralSecurityException("Exception to test", illegalStateException));
+    cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT, DEFAULT_CRYPTO_JOB_TIMEOUT_MS, mockTime);
     SecretKeySpec perBlobSecretKey = kms.getRandomKey();
-    testFailureOnDecryption(perBlobSecretKey, null, false, mockCryptoService, kms);
+    testFailureOnDecryption(perBlobSecretKey, null, false, mockCryptoService, kms, illegalStateException, false, -1);
     mockCryptoService.clearStates();
     cryptoJobHandler.close();
     MockKeyManagementService mockKms = new MockKeyManagementService(new KMSConfig(verifiableProperties), defaultKey);
-    cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT);
-    testFailureOnDecryption(perBlobSecretKey, mockKms, true, cryptoService, mockKms);
+    cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT, DEFAULT_CRYPTO_JOB_TIMEOUT_MS, mockTime);
+    testFailureOnDecryption(perBlobSecretKey, mockKms, true, cryptoService, mockKms, illegalStateException, false, -1);
+    cryptoJobHandler.close();
+
+    // test timeouts
+    long[] timeToBlockKms =
+        {DEFAULT_CRYPTO_JOB_TIMEOUT_MS + 2, DEFAULT_CRYPTO_JOB_TIMEOUT_MS * 2, DEFAULT_CRYPTO_JOB_TIMEOUT_MS * 5};
+    TimeoutException timeoutException = new TimeoutException();
+    for (long timeToBlock : timeToBlockKms) {
+      mockKms = new MockKeyManagementService(new KMSConfig(verifiableProperties), defaultKey);
+      cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT, DEFAULT_CRYPTO_JOB_TIMEOUT_MS, mockTime);
+      testFailureOnDecryption(perBlobSecretKey, mockKms, false, cryptoService, mockKms, timeoutException, true,
+          timeToBlock);
+      cryptoJobHandler.close();
```

Comment:
do you have to do the `clearStates()` call here? you do it in the encryption equivalent.